parse/planner/optimizer/pushdown.go

```diff
@@ -0,0 +1,236 @@
+				if _, ok := cols[[2]string{field.Parent, field.Name}]; ok {
+					rightCount++
+				}
+			}
+
+			switch {
+			case leftCount == 0 && rightCount == 0:
+				// we can't push down the filter
+				leftover = makeAnd(leftover, and)
+			case leftCount == 0 && rightCount > 0:
+				// push down to the right side
+				res, err := push(right, and)
+				if err != nil {
+					return nil, nil, nil
+				}
+
+				right = res
+			case leftCount > 0 && rightCount == 0:
+				// push down to the left side
+				res, err := push(left, and)
+				if err != nil {
+					return nil, nil, nil
+				}
+
+				left = res
+			case leftCount > 0 && rightCount > 0:
+				// apply the filter to the join condition
+				leftover = makeAnd(leftover, and)
+			default:
+				panic("unexpected column count case")
+			}
+		}
+
+		return left, right, leftover
+	}
+
+	switch n := n.(type) {
+	case *logical.Filter:
+		if expr != nil {
+			return nil, fmt.Errorf("unexpected pushdown of filter to filter")
+		}
+
+		if _, ok := n.Child.(*logical.Aggregate); ok {
+			// we can't push down filters to aggregates
+			return n, nil
+		}
+
+		fin, err := push(n.Child, n.Condition)
+		if err != nil {
+			return nil, err
+		}
+
+		// since we no longer have a condition, we can just return the child
+		return fin, nil
+	case *logical.Join:
+		left, right, leftover := pushLeftRight(n.Left, n.Right, expr)
+		n.Left = left
+		n.Right = right
+		n.Condition = makeAnd(n.Condition, leftover)
+
+		return n, nil
+	case *logical.Scan:
+		n.Filter = makeAnd(n.Filter, expr)
+		return n, nil
+	case *logical.Project:
+		res, err := push(n.Child, expr)
+		if err != nil {
+			return nil, err
+		}
+
+		n.Child = res
+		return n, nil
+	case *logical.CartesianProduct:
+		if expr == nil {
+			return n, nil
+		}
+
+		left, right, leftover := pushLeftRight(n.Left, n.Right, expr)
+		n.Left = left
+		n.Right = right
+		// if leftover is not nil, then we can rewrite as a join
+		if leftover == nil {
+			return n, nil
+		}
+
+		return &logical.Join{
+			Left:      left,
+			Right:     right,
+			Condition: leftover,
+		}, nil
+	case *logical.Update:
+		res, err := push(n.Child, expr)
+		if err != nil {
+			return nil, err
+		}
+
+		n.Child = res
+		return n, nil
+	case *logical.Delete:
+		res, err := push(n.Child, expr)
+		if err != nil {
+			return nil, err
+		}
+
+		n.Child = res
+		return n, nil
+	case *logical.Aggregate:
+		res, err := push(n.Child, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		n.Child = res
+		return n, nil
+	default:
+		if expr != nil {
+			return nil, fmt.Errorf("unhandled predicate pushdown for %T", n)
+		}
+
+		// by default, we just can't push down the filter, but we should visit the children.
+		// If any rewrite is attempted, we should return an error.
+		for _, child := range n.Plans() {
+			res, err := push(child, nil)
+			if err != nil {
+				return nil, err
+			}
+			if res != child {
+				return nil, fmt.Errorf("unhandled rewrite: tried to rewrite a %T as a child of a %T", res, n)
```

Comment:
No, it shouldn't. If the `push` call here changed the child, it wouldn't actually change the plan, because we don't re-assign the child anywhere. This is why, for example, we have to have a case for `Aggregate`, even though there is no logic:

```go
	case *logical.Aggregate:
		res, err := push(n.Child, nil)
		if err != nil {
			return nil, err
		}

		n.Child = res
		return n, nil
```

We can only rewrite a node if we are re-assigning it. In the `default` case, if the node is changed, we don't know how to rewrite it, and so it means that we generated a plan that we do not know how to optimize.

Reply:
Let's say we got a `Limit` plan, so `Limit -> Filter -> scan`, after push,  `Limit`'s child `Filter` should changed, right?

Reply:
Just to repost what we just discussed, `Limit -> Filter -> scan` is not possible in this planner.  
The implied combinations should be documented somewhere IMO.